File: Graph/859/kruskal.cpp
#include <iostream>
#include <cstring>
#include <algorithm>
using namespace std;

const int N = 2e5 + 10;
int n, m, x, y, z, a, b, pa, pb, res, cnt, p[N];
struct Edge{
    int a, b, c;
    bool operator <(const Edge &edge) {
        return c < edge.c;
    }
} e[N];

int find(int x);

int main() {
    scanf("%d%d", &n, &m);
    for (int i = 1; i <= n; i++) p[i] = i;
    for (int i = 0; i < m; i++) {
        scanf("%d%d%d", &x, &y, &z);
        e[i] = {x, y, z};
    }
    sort(e, e + m);
    for (int i = 0; i < m; i++) {
        a = e[i].a, b = e[i].b;
        pa = find(a), pb = find(b);
        if (pa != pb) {
            p[pa] = pb;
            res += e[i].c;
            cnt++;
        }
    }
    if (cnt < n - 1) printf("impossible");
    else printf("%d", res);
    return 0;
}

int find(int x) {
    if (p[x] != x) p[x] = find(p[x]);
    return p[x];
}

File: Graph/859/kruskal.py
def find(x: int, p: list) -> None:
    root = x
    while root != p[root]:
        root = p[root]
    while x != p[x]:
        t = p[x]
        p[x] = root
        x = t
    return root

if __name__ == "__main__":
    n, m = map(int, input().split())
    p = [i for i in range(n + 1)]
    e = [tuple(map(int, input().split())) for _ in range(m)]
    e.sort(key=lambda x: x[2])
    res, cnt = 0, 0
    for x, y, z in e:
        px, py = find(x, p), find(y, p)
        if px != py:
            p[px] = py
            cnt += 1
            res += z
    if cnt < n - 1:
        print("impossible")
    else:
        print(res)
